### PR TITLE
Update `single_bead.py` to include thermal history changes

### PR DESCRIPTION
**Pull request by safeerehman**
- Update class `MeltPool` to include thermal history output
- Update class `SingleBeadInput` to include additional required args
- Update relevant test cases
- Update single bead example to show the usage
- The end result is a gif file that gets generated for the example -
- 
image

### CODE REVIEW

File: src/ansys/additive/core/single_bead.py [hunk 1]
```diff
@@ -193,8 +244,36 @@ def __eq__(self, __o: object) -> bool:
...
+        if value is None:
+            self._thermal_history_output = None
+            return
+        self._thermal_history_output = pathlib.Path(value)
+        if not os.path.exists(self._thermal_history_output):
+            os.makedirs(self._thermal_history_output)
+
+    def _unarchive_grid_full_thermal_sensor_file(self, thermal_history_vtk: bytes):
+        """Unarchive the grid full thermal sensor file output."""
+        self.thermal_history_output = self.CONST_GRID_FULL_THERMAL_HISTORY_OUTPUT_PATH
```
Comment (pkrull-ansys) [inline src/ansys/additive/core/single_bead.py:269]: This path is wrong. It needs to be <user data path>/<simulation id>/CONST_GRID_FULL_THERMAL_HISTORY_OUTPUT_PATH. See what we do in `_simulate_thermal_history()`. You'll want to add `out_dir` to the parameter list for `SingleBeadSummary.__init__()`, just like `ThermalHistorySummary.__init__()`. In `simulate()`, you can set the `out_dir` value to `USER_DATA_PATH`.

With what you have, if the user ran two single bead simulations with thermal-history, the second simulation would overwrite the thermal history results from the first one.

**Review by pkrull-ansys — CHANGES_REQUESTED** [review]

File: src/ansys/additive/core/single_bead.py [hunk 2]
```diff
@@ -120,12 +137,38 @@ def bead_length(self, value):
         self.__validate_range(value, self.MIN_BEAD_LENGTH, self.MAX_BEAD_LENGTH, "bead_length")
         self._bead_length = value
 
+    @property
+    def output_thermal_history(self) -> bool:
+        """Flag to output the thermal history of the simulation."""
```
Comment (julieatansys) [inline src/ansys/additive/core/single_bead.py:142]:
```suggestion
        """Option to output the thermal history of the simulation."""
```

File: src/ansys/additive/core/single_bead.py [hunk 3]
```diff
@@ -193,8 +244,36 @@ def __eq__(self, __o: object) -> bool:
...
+        """Set the path to the VTK file(s) containing the thermal history of the simulation."""
+        if value is None:
+            self._thermal_history_output = None
+            return
+        self._thermal_history_output = pathlib.Path(value)
+        if not os.path.exists(self._thermal_history_output):
+            os.makedirs(self._thermal_history_output)
+
+    def _unarchive_grid_full_thermal_sensor_file(self, thermal_history_vtk: bytes):
+        """Unarchive the grid full thermal sensor file output."""
```
Comment (julieatansys) [inline src/ansys/additive/core/single_bead.py:268]:
```suggestion
        """Unarchive the grid full thermal sensor file output."""
        I don't know what you mean by this. It is not clear.
```

Comment (julieatansys) [inline src/ansys/additive/core/single_bead.py:268]: I don't know what you mean by this. Not clear.

File: src/ansys/additive/core/single_bead.py (hunk 2 above)
Comment (PipKat) [reply src/ansys/additive/core/single_bead.py:142]: For a property, the format is generally: "Flag indicating whether to output the thermal history of the simulation."

Comment (julieatansys) [reply src/ansys/additive/core/single_bead.py:142]: OK good, go with that! Thanks, Kathy.

File: src/ansys/additive/core/single_bead.py [hunk 4]
```diff
@@ -40,19 +43,31 @@ class SingleBeadInput:
...
     MAX_BEAD_LENGTH = 1e-2
     """Maximum bead length (m)."""
+    DEFAULT_OUTPUT_THERMAL_HISTORY = False
+    """Default output thermal history flag."""
+    DEFAULT_THERMAL_HISTORY_INTERVAL = 1
+    """Default thermal history interval."""
+    MIN_THERMAL_HISTORY_INTERVAL = 1
+    """Minimum thermal history interval."""
+    MAX_THERMAL_HISTORY_INTERVAL = 10000
+    """Maximum thermal history interval."""
```
[the remark this answers is not in the document]
Comment (safeerehman) [reply src/ansys/additive/core/single_bead.py:54]: Works fine the `download_file()` method.

File: src/ansys/additive/core/additive.py [hunk 5]
```diff
@@ -402,7 +402,7 @@ def _simulate(
                     if progress.state == ProgressState.ERROR:
                         raise Exception(progress.message)
                 if response.HasField("melt_pool"):
-                    return SingleBeadSummary(input, response.melt_pool)
+                    return self._check_for_thermal_history(response, input)
```
Comment (pkrull-ansys) [inline src/ansys/additive/core/additive.py:405]: There's a design smell here. `check_for_*` should either return a bool or it should return void upon success and throw an exception if the check fails. There are better ways to do what you're trying to do. Here's one approach. First update the protobuf definitions to add `optional` to the field.
```
message MeltPool {
    repeated MeltPoolTimeStep time_steps = 1;
    optional string thermal_history_vtk_zip = 2;
}
```
Then, at this point in the code, add `if response.meltpool.has_thermal_history_vtk_zip:` followed by the download of the zip file. After this, call `return SingleBeadSummary(input, response.melt_pool, path_to_zip_file)` and pass in the zip file, NOT the folder the contents were extracted to. Do the extraction in the `init()` method of `SingleBeadSummary`.

File: src/ansys/additive/core/additive.py [hunk 6]
```diff
@@ -624,3 +624,26 @@ def _check_for_duplicate_id(self, inputs):
             if input.id in ids:
                 raise ValueError(f'Duplicate simulation ID "{input.id}" in input list')
             ids.append(input.id)
+
+    def _check_for_thermal_history(self, response, input) -> SingleBeadSummary:
```
Comment (pkrull-ansys) [inline src/ansys/additive/core/additive.py:628]: The bulk of this should be moved to `__init__()` in `SingleBeadSummary` as mentioned above.

**Review by pkrull-ansys — CHANGES_REQUESTED** [review]

File: src/ansys/additive/core/additive.py (hunk 5 above)
Comment (safeerehman) [reply src/ansys/additive/core/additive.py:405]: Okay, I see that the protobuf definitions are optional by default. So will try to organize the code in a way that the `additive.py` just downloads the zip file, but the `SingleBeadSummary` takes care of extracting the contents.

**Review by pkrull-ansys — APPROVED** [review]
I'm approving this so you can merge it while I'm gone. Please implement the previously suggested changes before merging.

**Comment by pkrull-ansys** [comment]
Safee,
Look at https://github.com/protocolbuffers/protobuf/blob/v3.12.0/docs/field_presence.md#considerations-for-change-compatibility. Adding optional to the field definition may be unnecessary but it's not the same as leaving it off. After reading this, you can decide which method to use.
[https://opengraph.githubassets.com/31d23c01779651bcfc23b075bc1495f827ebc7949e75af64509bcac949457e8b/protocolbuffers/protobuf]<https://github.com/protocolbuffers/protobuf/blob/v3.12.0/docs/field_presence.md#considerations-for-change-compatibility>
protobuf/docs/field_presence.md at v3.12.0 · protocolbuffers/protobuf<https://github.com/protocolbuffers/protobuf/blob/v3.12.0/docs/field_presence.md#considerations-for-change-compatibility>
Protocol Buffers - Google's data interchange format - protocolbuffers/protobuf
github.com



Peter Krull

435 258-8202



________________________________

**Pull request merged**